ostd/src/sync/rcu/mod.rs

```diff
@@ -2,105 +2,406 @@
 }
 
-#[repr(transparent)]
-pub struct RcuReclaimer<P> {
-    ptr: P,
+// RCU guards that have a non-null pointer can be directly dereferenced.
+impl<P: OwnerPtr + Send> Deref for RcuReadGuard<'_, P> {
+    type Target = <P as OwnerPtr>::Target;
+
+    fn deref(&self) -> &Self::Target {
+        // SAFETY:
+        // 1. This pointer is not NULL because the type is `RcuReadGuard`.
+        // 2. Since the preemption is disabled, the pointer is valid because
+        //    other writers won't release the allocation until this task passes
+        //    the quiescent state.
+        unsafe { &*self.0.obj_ptr }
+    }
 }
 
-impl<P: Send + 'static> RcuReclaimer<P> {
-    pub fn delay(mut self) {
-        let ptr: P = unsafe {
-            let ptr = core::mem::replace(&mut self.ptr, core::mem::uninitialized());
+// RCU guards that may have a null pointer can be dereferenced after checking.
+impl<P: OwnerPtr + Send> RcuOptionReadGuard<'_, P> {
+    /// Gets the reference of the protected data.
+    ///
+    /// If the RCU cell contains none, this function returns `None`.
+    pub fn get(&self) -> Option<&<P as OwnerPtr>::Target> {
+        if self.0.obj_ptr.is_null() {
+            return None;
+        }
+        // SAFETY:
+        // 1. This pointer is not NULL.
+        // 2. Since the preemption is disabled, the pointer is valid because
+        //    other writers won't release the allocation until this task passes
+        //    the quiescent state.
+        Some(unsafe { &*self.0.obj_ptr })
+    }
 
-            core::mem::forget(self);
+    /// Returns if the RCU cell contains none when doing [`Rcu::read`].
+    pub fn is_none(&self) -> bool {
+        self.0.obj_ptr.is_null()
+    }
 
-            ptr
-        };
-        get_singleton().after_grace_period(move || {
-            drop(ptr);
-        });
+    /// Tries to replace the already read pointer with a new pointer
+    /// (or none).
+    ///
+    /// If another thread has updated the pointer after the read, this
+    /// function will fail, and returns the given pointer back. Otherwise,
+    /// it will replace the pointer with the new one and drop the old pointer
+    /// after the grace period.
+    ///
+    /// If spinning on [`RcuOption::read`] and this function, it is recommended
+    /// to relax the CPU or yield the task on failure. Otherwise contention
+    /// will occur.
+    ///
+    /// This API does not help to avoid
+    /// [the ABA problem](https://en.wikipedia.org/wiki/ABA_problem).
+    pub fn compare_exchange(self, new_ptr: Option<P>) -> Result<(), Option<P>> {
+        self.0.compare_exchange(new_ptr)
     }
 }
 
-impl<P> Drop for RcuReclaimer<P> {
-    fn drop(&mut self) {
-        let wq = Arc::new(WaitQueue::new());
-        get_singleton().after_grace_period({
-            let wq = wq.clone();
-            move || {
-                wq.wake_one();
-            }
-        });
-        wq.wait_until(|| Some(0u8));
+impl<P: OwnerPtr + Send> RcuReadGuard<'_, P> {
```

Comment:
Could you please move this `impl` just below `impl<P: OwnerPtr + Send> Deref for RcuReadGuard<'_, P>`?

Then we'll have `impl` blocks ordered like `impl Rcu`, `impl RcuOption`, `impl RcuReadGuard`, `impl RcuOptionReadGuard`, which sounds cleaner.